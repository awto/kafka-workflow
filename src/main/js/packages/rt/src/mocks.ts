/**
 * lightweight workflow's testing
 *
 * the workflow functions are usual JS async functions, and can be debugged and tested as any other JS function
 */
export const config = {
  outputTopics: new Set(["workflow-result", "workflow-error"]),
  resultTopic: "workflow-result",
  errorTopic: "workflow-error"
};

export let threadId = "testThread";
export let stepId = "testStep";

export const __topicHandlers = new Map<
  string,
  (msg: string, dest: string, topic: string) => unknown
>();

interface Output {
  msg: string;
  dest: string;
  topic: string;
}

export const __trace: any[] = [];
export const __suspended = new Map<string, Suspended>();
export let __lastTopicUpdate: {[topicName:string]:{k:string,v:string}} = {};

export async function __resume(id: string, value: unknown) {
  __trace.push({resume:id, value});
  await new Promise(i => setTimeout(i, 0))
  const suspended = __suspended.get(id);
  if (!suspended) throw new Error(`nothing is suspended with id ${id}`);
  __suspended.delete(id);
  for (const i of suspended.onfulfilled) i(value);
  await new Promise(i => setTimeout(i, 0))
}

export async function __reject(id: string, reason: unknown) {
  await new Promise(i => setTimeout(i, 10))
  __trace.push({reject:id, reason});
  const suspended = __suspended.get(id);
  if (!suspended) throw new Error(`nothing is suspended with id ${id}`);
  __suspended.delete(id);
  for (const i of suspended.onrejected) i(reason);
  // suspended.onrejected && suspended.onrejected(reason);
  await new Promise(i => setTimeout(i, 10))
}

/** resets state in tests setup */
export function __reset() {
  __lastTopicUpdate = {}
  __trace.length = 0;
  Suspended.count = 0;
  __topicHandlers.clear();
}

export class Suspended {
  static count = 0;
  onfulfilled: ((arg: unknown) => unknown)[] = [];
  onrejected: ((arg: unknown) => unknown)[] = [];
  constructor(public id: string = `${stepId}:${++Suspended.count}`) {
    __suspended.set(id, this);
  }
  then(onff: (arg: unknown) => unknown, onrj?: (arg: unknown) => unknown) {
    this.onfulfilled.push(onff);
    if (onrj) this.onrejected.push(onrj);
  }
}

export function suspend(id?: string): any {
  return new Suspended(id);
}

export function output(msg: string, topic: string, dest = threadId) {
  const item = { dest, msg, topic };
  __trace.push(item);
  const handler = __topicHandlers.get(topic);
  __lastTopicUpdate[topic] = {k:dest, v:msg};
  if (handler) handler(msg, dest, topic);
}

export function outputJSON(msg: unknown, topic: string, dest = threadId) {
  return output(JSON.stringify(msg), topic, dest);
}

export const any = (arr:any[]) => Promise.race(arr);
export const all = (arr:any[]) => Promise.all(arr);
export const anyWithCancel = any;
export const allWithCancel = all;
