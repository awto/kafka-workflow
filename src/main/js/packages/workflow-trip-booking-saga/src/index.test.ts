import * as W from "@effectful/kafka-workflow/lib/mocks";
import * as Index from "./index";

afterEach(() => {
  W.__reset();
});

async function resumeResource(kind: string) {
  await W.__resume(W.__lastTopicUpdate[`saga-reserve-${kind}`].v, {
    id: `${kind}-id`
  });
}

test("concurent execution", async function () {
  const res = Index.main();
  await resumeResource("flight");
  await resumeResource("hotel");
  await resumeResource("car");
  const item = await res;
  expect(item).toEqual({
    car: {
      id: "testStep:1"
    },
    flight: {
      id: "testStep:3"
    },
    hotel: {
      id: "testStep:2"
    }
  });
  expect(W.__trace).toMatchSnapshot();
  expect(W.__suspended.has("main")).toBe(false);
});

test("one thread is rejected", async function () {
  const res = Index.main();
  expect(res).rejects.toBe("not available")
  await resumeResource("flight");
  await W.__reject(
    W.__lastTopicUpdate[`saga-reserve-hotel`].v,
    "not available"
  );
  expect(W.__trace).toMatchSnapshot();
  expect(W.__suspended.has("main")).toBe(false);
});

test("timeout", async function () {
  const res = Index.main();
  expect(res).rejects.toBe("timeout");
  await resumeResource("car");
  await W.__resume(
    JSON.parse(W.__lastTopicUpdate[`workflow-scheduler`].k.split("|")[1]).ref,
    undefined
  );
  expect(W.__trace).toMatchSnapshot();
  expect(W.__suspended.has("main")).toBe(false);
});
