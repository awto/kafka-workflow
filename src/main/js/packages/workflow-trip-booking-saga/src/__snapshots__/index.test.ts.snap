// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`concurent execution 1`] = `
Array [
  Object {
    "dest": "testThread",
    "msg": "testStep:1",
    "topic": "saga-reserve-car",
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:2",
    "topic": "saga-reserve-hotel",
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:3",
    "topic": "saga-reserve-flight",
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:4\\"}",
    "msg": "1000",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "testStep:3",
    "value": Object {
      "id": "flight-id",
    },
  },
  Object {
    "resume": "testStep:2",
    "value": Object {
      "id": "hotel-id",
    },
  },
  Object {
    "resume": "testStep:1",
    "value": Object {
      "id": "car-id",
    },
  },
]
`;

exports[`one thread is rejected 1`] = `
Array [
  Object {
    "dest": "testThread",
    "msg": "testStep:1",
    "topic": "saga-reserve-car",
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:2",
    "topic": "saga-reserve-hotel",
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:3",
    "topic": "saga-reserve-flight",
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:4\\"}",
    "msg": "1000",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "testStep:3",
    "value": Object {
      "id": "flight-id",
    },
  },
  Object {
    "reason": "not available",
    "reject": "testStep:2",
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:3",
    "topic": "saga-cancel-flight",
  },
]
`;

exports[`timeout 1`] = `
Array [
  Object {
    "dest": "testThread",
    "msg": "testStep:1",
    "topic": "saga-reserve-car",
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:2",
    "topic": "saga-reserve-hotel",
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:3",
    "topic": "saga-reserve-flight",
  },
  Object {
    "dest": "testThread|{\\"ref\\":\\"testStep:4\\"}",
    "msg": "1000",
    "topic": "workflow-scheduler",
  },
  Object {
    "resume": "testStep:1",
    "value": Object {
      "id": "car-id",
    },
  },
  Object {
    "resume": "testStep:4",
    "value": undefined,
  },
  Object {
    "dest": "testThread",
    "msg": "testStep:1",
    "topic": "saga-cancel-car",
  },
]
`;
